processor/otel/metrics.go

```diff
@@ -0,0 +1,235 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Portions copied from OpenTelemetry Collector (contrib), from the
+// elastic exporter.
+//
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otel
+
+import (
+	"context"
+	"sort"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/publish"
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// ConsumeMetrics consumes OpenTelemetry metrics data, converting into
+// the Elastic APM metrics model and sending to the reporter.
+func (c *Consumer) ConsumeMetrics(ctx context.Context, metrics pdata.Metrics) error {
+	batch := c.convertMetrics(metrics)
+	return c.Reporter(ctx, publish.PendingReq{
+		Transformables: batch.Transformables(),
+		Trace:          true,
+	})
+}
+
+func (c *Consumer) convertMetrics(metrics pdata.Metrics) *model.Batch {
+	batch := model.Batch{}
+	resourceMetrics := metrics.ResourceMetrics()
+	for i := 0; i < resourceMetrics.Len(); i++ {
+		c.convertResourceMetrics(resourceMetrics.At(i), &batch)
+	}
+	return &batch
+}
+
+func (c *Consumer) convertResourceMetrics(resourceMetrics pdata.ResourceMetrics, out *model.Batch) {
+	var metadata model.Metadata
+	translateResourceMetadata(resourceMetrics.Resource(), &metadata)
+	instrumentationLibraryMetrics := resourceMetrics.InstrumentationLibraryMetrics()
+	for i := 0; i < instrumentationLibraryMetrics.Len(); i++ {
+		c.convertInstrumentationLibraryMetrics(instrumentationLibraryMetrics.At(i), metadata, out)
+	}
+}
+
+func (c *Consumer) convertInstrumentationLibraryMetrics(in pdata.InstrumentationLibraryMetrics, metadata model.Metadata, out *model.Batch) {
+	var ms metricsets
+	otelMetrics := in.Metrics()
+	var unsupported int64
+	for i := 0; i < otelMetrics.Len(); i++ {
+		if !c.addMetric(otelMetrics.At(i), &ms) {
```

Comment:
instead of having a method that both mutates and argument and returns a value, can't we say `unsupported += otelMetrics.Len() - ms.Len()`?

Reply:
No, because two metrics may end up in the same metricset. (`ms` is a collection of metricsets)